### PR TITLE
[tensorflow] make SourceKit use an inmemory filesystem for clang module cache

### PR DESCRIPTION
**Pull request by marcrasi**
Context about the problem: We'd like to run SourceKit on a server where we're not allowed to write to the filesystem. The ClangImporter writes module caches to the filesystem, so we need some way to stop that from happening.

This PR solves the problem by using the `InMemoryOutputFileSystem` defined in https://github.com/apple/swift-clang/pull/330.

This may only be a temporary solution, but everything else I tried was quite hard to get working (explicitly building required clang modules before invoking SourceKit, so that SourceKit doesn't have to build the modules into the cache), so I'd like to try this for a while.

### CODE REVIEW

File: test/SourceKit/CursorInfo/in_memory_clang_module_cache.swift [hunk 1]
```diff
@@ -0,0 +1,13 @@
+let foo: Int = 10
+foo
+
+// Checks that the SourceKit request succeeded.
+// CHECK-SOURCEKIT: source.lang.swift.ref.var.global (1:5-1:8)
+
+// Checks that nothing has been written into the module cache on the real
+// filesyustem.
```
Comment (burmako) [inline test/SourceKit/CursorInfo/in_memory_clang_module_cache.swift]:
```suggestion
// filesystem.
```

File: tools/SourceKit/include/SourceKit/Core/LangSupport.h [hunk 2]
```diff
@@ -638,6 +640,12 @@ class LangSupport {
 
   virtual ~LangSupport() { }
 
+  // SWIFT_ENABLE_TENSORFLOW
+  /// Subseqent requests will write temporary output files to this filesystem
```
Comment (burmako) [inline tools/SourceKit/include/SourceKit/Core/LangSupport.h]:
```suggestion
  /// Subsequent requests will write temporary output files to this filesystem
```

File: tools/SourceKit/lib/SwiftLang/SwiftASTManager.h [hunk 3]
```diff
@@ -93,6 +95,12 @@ class SwiftASTManager : public std::enable_shared_from_this<SwiftASTManager> {
                            StringRef RuntimeResourcePath);
   ~SwiftASTManager();
 
+  // SWIFT_ENABLE_TENSORFLOW
+  /// Subseqent requests will write temporary output files to this filesystem
```
Comment (burmako) [inline tools/SourceKit/lib/SwiftLang/SwiftASTManager.h]:
```suggestion
  /// Subsequent requests will write temporary output files to this filesystem
```

File: tools/SourceKit/tools/sourcekitd-test/Options.td [hunk 4]
```diff
@@ -137,6 +137,10 @@ def repeat_request_EQ : Joined<["-"], "repeat-request=">, Alias<repeat_request>;
 def vfs_files : CommaJoined<["-"], "vfs-files=">,
   HelpText<"Injects a VFS into the request, overlaying files specified by the given <name>=<target> pairs over the real filesystem">;
 
+// SWIFT_ENABLE_TENSORFLOW
+def in_memory_clang_module_cache : Flag<["-"], "in-memory-clang-module-cache">,
+  HelpText<"Put the clang module cache in memory">;
```
Comment (burmako) [inline tools/SourceKit/tools/sourcekitd-test/Options.td]:
```suggestion
  HelpText<"Put the Clang module cache in memory">;
```

File: tools/SourceKit/tools/sourcekitd-test/TestOptions.cpp [hunk 5]
```diff
@@ -367,6 +367,17 @@ bool TestOptions::parseArgs(llvm::ArrayRef<const char *> Args) {
...
 #endif
 
+    // SWIFT_ENABLE_TENSORFLOW
+    case OPT_in_memory_clang_module_cache:
+#ifdef SWIFT_SOURCEKIT_USE_INPROC_LIBRARY
+      InMemoryClangModuleCache = true;
+      break;
+#else
+      llvm::errs() << "in-memory-clang-module-cache only supported when "
+                      "SWIFT_SOURCEKIT_USE_INPROC_LIBRARY is set";
```
Comment (burmako) [inline tools/SourceKit/tools/sourcekitd-test/TestOptions.cpp]: Could you elaborate on the reasons behind this restriction?

File: tools/SourceKit/tools/sourcekitd/include/sourcekitd/FileSystemProvider.h [hunk 6]
```diff
@@ -31,6 +33,16 @@ namespace SourceKit {
 void setGlobalFileSystemProvider(
     llvm::StringRef Name, SourceKit::FileSystemProvider *FileSystemProvider);
 
+/// Subseqent requests will write temporary output files to this filesystem
```
Comment (burmako) [inline tools/SourceKit/tools/sourcekitd/include/sourcekitd/FileSystemProvider.h]:
```suggestion
/// Subsequent requests will write temporary output files to this filesystem
```

**Review by burmako — APPROVED** [review]

File: tools/SourceKit/tools/sourcekitd-test/TestOptions.cpp (hunk 5 above)
Comment (marcrasi) [reply tools/SourceKit/tools/sourcekitd-test/TestOptions.cpp]: added comment

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Pull request merged**